### DISCUSSION TITLE
ch6/ch6-04

### DISCUSSION THREAD

**Discussion by giscus** [gopl-zh/gopl-zh.github.com#162] [General]
# ch6/ch6-04

<The Go Programming Language>中文版

https://gopl-zh.github.io/ch6/ch6-04.html

**Comment by zhangbest5** [comment]
真是罗里吧嗦，简单直接用对比不好么？

**Comment by erich5u** [comment]
这么简单的东西能说的这么绕。。

**Comment by WYXNB** [comment]
这段文本讲解了方法值（method values）和方法表达式（method expressions）的概念及其在 Go 编程语言中的应用。让我来为你进行简要的总结：

方法值（Method Values）：
在 Go 中，方法值是将一个方法绑定到特定接收器对象上并得到一个函数值。通过这个函数值，你可以在不需要指定接收器的情况下调用该方法。方法值的创建类似于函数闭包，它捕获了接收器作为隐藏参数。方法值的创建可以通过将方法名赋值给变量来完成，然后这个变量就可以作为函数调用。

示例：

go
Copy code
type Point struct{ X, Y float64 }

func (p Point) Distance(q Point) float64 {
    // calculate distance between p and q
}

p := Point{1, 2}
q := Point{4, 6}
distanceFromP := p.Distance // method value
distance := distanceFromP(q) // Call the method using method value
方法表达式（Method Expressions）：
方法表达式允许你直接调用一个方法，而不需要一个接收器实例。通过类型名称加上方法名的方式，你可以获得一个函数值，这个函数值可以接收实际的接收器作为第一个参数，从而调用方法。方法表达式的语法允许你在调用时提供接收器作为参数，而不需要提前创建一个接收器对象。

示例：

go
Copy code
type Point struct{ X, Y float64 }

func (p Point) Distance(q Point) float64 {
    // calculate distance between p and q
}

distanceFunc := Point.Distance // method expression
distance := distanceFunc(p, q) // Call the method using method expression
这些概念在实际编程中很有用，特别是当你需要在某些情况下动态地选择要调用的方法时，例如在根据不同条件选择不同方法执行时。这种技术在提高代码的灵活性和可重用性方面非常有帮助。

**Comment by NotFool** [comment]
方法这一章写的好差，绕而且无意义的在重复。

**Comment by aahowe** [comment]
随便问个大模型比他讲的清晰

**Comment by 2466564159** [comment]
还好不是我一个人觉得好绕。。。

**Comment by Zacharia2** [comment]
- 选择器（p.Distance）->p对象的方法值，调用方式是：方法值存入的变量名（方法值的参数）
- 方法表达式（Point.Distance）->Point类型的函数值，调用方式是：函数值存入的变量名（指定接收器，函数值的参数）
